### ISSUE TITLE
Site Editor: Full width Cover blocks that use Featured Image show a visual glitch when resizing in the preview pane

### ISSUE THREAD

**Issue by jordesign** [WordPress/gutenberg#52313]
### Description

When a Template or Pattern uses a Full width cover block that is set to use the featured image - it shows a strange flicker in the placeholder/background when the preview panel resizes (on hover).

### Step-by-step reproduction instructions

1. Open the site editor and edit/add a template.
2. Add a Cover block, set to use the Featured Image as the background.
3. Set the block to Full width.
4. Save and close the template - returning to the site editor browse/mode.
5. Hover over the preview of the template.

### Screenshots, screen recording, code snippet

In a pattern - showing it only happens with full-width aligned blocks.

https://github.com/WordPress/gutenberg/assets/1842363/0bcaf4e2-b290-4d53-a84d-fef1ef29ce47

In a template

https://github.com/WordPress/gutenberg/assets/1842363/d5043a66-ba7d-4527-a5e7-589c3cdc7672

### Environment info

- WordPress 6.3 beta3
- Gutenberg 16.1

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by carolinan** [comment]
I am able to reproduce this on macOs Chrome Version 114.0.5735.198  but not in Firefox 115.0 or Safari 16.5.1 (18615.2.9.11.7)

**Comment by ndiego** [comment]
I'm struggling to reproduce this in Chrome 🤔

**Comment by jordesign** [comment]
Version 114.0.5735.133 of Chrome for me

**Comment by carolinan** [comment]
PC Windows 10, Chrome Version 114.0.5735.199, WP 6.3-beta3-56143:
Without Gutenberg: No issue
With Gutenberg Version 16.1.1: No issue
With Gutenberg trunk: No issue

It makes no difference if I try changing the browser window width or screen resolution on either the PC or Mac.

If I view a page instead of a template, I do not see the flickering.

**Comment by jordesign** [comment]
Noting this is still in place in RC1

**Comment by annezazu** [comment]
Noting this can be replicated with 6.3.1

**Comment by glendaviesnz** [comment]
Looks like there is a [known issue in chrome with flickering](https://www.flowradar.com/answer/how-to-reduce-flickering-caused-by-blurred-items-with-large-backdrop-filters-on-chrome-when-using-webflow) when blur backdrop filter is applied, which is the case on the featured image placeholder. There are some options for fixing this in the attached blog post that I will take a look at.

**Comment by glendaviesnz** [comment]
@jasmussen It looks like it was [the blur that was added here](https://github.com/WordPress/gutenberg/pull/43379/files#diff-e6d39e2bf2050a2e7a7022d39b3ddb498bf87165293fd8759302bab4c2dfb346R183) that causes this. I have had a look and can't see any way to stop the flicker in this context. Do you have any thoughts on alternatives to get the same UI effect without the blur?

**Comment by jasmussen** [comment]
Fascinating! Thanks for the ping. 

This is why the blur was added, so that the placeholder would always be legible if placed on a background image:

<img width="795" alt="Screenshot 2023-09-26 at 10 03 02" src="https://github.com/WordPress/gutenberg/assets/1204802/8de879fa-f5c6-409e-92b0-855acf22b18d">

Curiously I'm not able to initially reproduce the original issue reported in Chrome. In fact, I'm seeing the backdrop-filter removed from placeholders through this rule:
<img width="352" alt="Screenshot 2023-09-26 at 09 59 36" src="https://github.com/WordPress/gutenberg/assets/1204802/a0c95796-3bc3-4f63-bf5b-7ac1e20d666c">

But if I manually check that off, I see it, sure enough:

![backdrop filter](https://github.com/WordPress/gutenberg/assets/1204802/d678928d-c8e2-4f77-8193-ff6d5d28ec61)

Curiously, I also see this weird behavior in the post editor, where if you place the caret in the tittle field, the caret itself gets blurred by the backdrop filter:

![curious](https://github.com/WordPress/gutenberg/assets/1204802/d8f79c60-21cd-472e-84c7-b4a0bff1858c)

Knowing that, and knowing that the frame scales on hover, it seems like the issue is the rounding math that is present when CSS `transform`s scale the preview frame. Shown in this GIF, how applying a border on this frame can color the blur that gets flickered in:

![it's the edge blurring in](https://github.com/WordPress/gutenberg/assets/1204802/0baa69b0-26d5-4b27-a3c3-669c43a41350)

That also kind of explains that the issue here is primarily browser _rounding math_, and the fact that _contents surrounding the blur filter can affect the inner contents_. The net result is somewhere between "to be expected" and "oh math".

So what do we do about fixes? Beyond any browser fixes, I've found two hacks that seem to address this, neither of which are great. First one is to add a little inner padding on the frame:

```
.block-editor-iframe__body {
	padding: 0.7px;
}
```

This is sub-optimal because it definitely is intentional that the frame has no inner padding. And 0.7px seems to be the lowest we can go to avoid the blur.

The second hack is probably a bit more acceptable since it affects just the placeholder state, but it's still not great since I believe it affects _all_ placeholder states. It's something along the lines of the following:

```
.wp-block-cover .wp-block-cover__image--placeholder-image {
    bottom: 0.8px;
    left: 0.8px;
    position: absolute;
    right: 0.8px;
    top: 0.8px;
    width: auto;
}
```

Here, we change the abs-positioning of the blurred illustration inside the placeholder to be _slightly pushed _inwards_ from the edge of the container. This ensures that the featured image doesn't ever touch the dark background edges, and is thus not needing to blur them when they bleed in. The code needs a little sanity check, and to live in the right place, but it could potentially be a heavily CSS commented hack for the time being.

That said, I wish there was a cleaner hack to contain the re-painting inside just the placeholder illustration and in case you're a CSS wizard inspired by the above (perhaps @t-hamano?) could find a better solution. Hope that helps!

**Comment by t-hamano** [comment]
Thank you for pinging me!

I was able to reproduce this issue on my Windows machine. It can only be reproduced in Chrome, and cannot be reproduced in Firefox.

<details><summary>Screencast in Chrome</summary>

https://github.com/WordPress/gutenberg/assets/54422211/9c62f788-a582-4b02-a096-4cbb8aab9ae7

</details> 

<details><summary>Screencast in Firefox</summary>

https://github.com/WordPress/gutenberg/assets/54422211/aff68634-d50e-4daa-9d8c-1ba3388ffb0b

</details> 

Here is the markup.

<details><summary>Details</summary>

```html
<!-- wp:cover {"useFeaturedImage":true,"dimRatio":50,"customOverlayColor":"#FFF","isUserOverlayColor":true,"isDark":false,"layout":{"type":"constrained"}} -->
<div class="wp-block-cover is-light"><span aria-hidden="true" class="wp-block-cover__background has-background-dim" style="background-color:#FFF"></span><div class="wp-block-cover__inner-container"><!-- wp:paragraph {"align":"center","placeholder":"Write title…","fontSize":"large"} -->
<p class="has-text-align-center has-large-font-size"></p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:cover -->

<!-- wp:cover {"url":"https://live.staticflickr.com/4572/38004374914_6b686d708e_b.jpg","id":1,"dimRatio":50,"customOverlayColor":"#FFF","isUserOverlayColor":true,"isDark":false,"align":"full","style":{"spacing":{"padding":{"top":"var:preset|spacing|80","bottom":"var:preset|spacing|80"}}},"layout":{"type":"constrained"}} -->
<div class="wp-block-cover alignfull is-light" style="padding-top:var(--wp--preset--spacing--80);padding-bottom:var(--wp--preset--spacing--80)"><span aria-hidden="true" class="wp-block-cover__background has-background-dim" style="background-color:#FFF"></span><img class="wp-block-cover__image-background wp-image-1" alt="" src="https://live.staticflickr.com/4572/38004374914_6b686d708e_b.jpg" data-object-fit="cover"/><div class="wp-block-cover__inner-container"><!-- wp:cover {"useFeaturedImage":true,"dimRatio":50,"customOverlayColor":"#FFF","isUserOverlayColor":true,"isDark":false,"layout":{"type":"constrained"}} -->
<div class="wp-block-cover is-light"><span aria-hidden="true" class="wp-block-cover__background has-background-dim" style="background-color:#FFF"></span><div class="wp-block-cover__inner-container"><!-- wp:paragraph {"align":"center","placeholder":"Write title…","fontSize":"large"} -->
<p class="has-text-align-center has-large-font-size"></p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:cover --></div></div>
<!-- /wp:cover -->
```

</details> 

I also tested a CSS hack solution, but it didn't work in my environment. It might also be affected by the size of the browser.

<details><summary>Screencast</summary>

https://github.com/WordPress/gutenberg/assets/54422211/b11fda45-71dc-487d-982f-2683fb9a63bf

</details> 

Another solution might be to apply a semi-transparent white background instead of `backdrop-filter`. In other words, CSS like below:

```css
.components-placeholder.has-illustration {
  background-color: rgba(255,255,255,0.5);
}
```

However, it will look different when placed in front of a background with colors or images.

| Before | After |
|--------|--------|
| ![before](https://github.com/WordPress/gutenberg/assets/54422211/f9e630a6-f7dc-4752-b7b4-e8a8b53c298c) | ![after](https://github.com/WordPress/gutenberg/assets/54422211/e7da04d3-dc55-4745-8327-d3759ee538a0) |

**Comment by jasmussen** [comment]
Thanks so much for looking. 

The main problem with the semitransparent background color is that if the background is noisy enough, it affects legibility.

Can you try a larger padding than 0.7px, just to see if that fixes it for you? Mainly as a test.

**Comment by t-hamano** [comment]
> Can you try a larger padding than 0.7px, just to see if that fixes it for you? Mainly as a test.

Unfortunately, the problem was not resolved in my environment.

<details><summary>Screencast</summary>

https://github.com/WordPress/gutenberg/assets/54422211/4132c1f3-f791-45fe-a0a5-d37e0191fa99

</details> 

In addition, I have tried several of the solutions in the article linked to in [this comment](https://github.com/WordPress/gutenberg/issues/52313#issuecomment-1728671104).

1. Optimize the backdrop filter: I tried lowering the blur value significantly, but could not eliminate the flickering completely.
2. Use hardware acceleration: In my environment, on the contrary, I was able to solve the problem by turning this setting **off**.

The following four would not be ideal solutions, as they would require design changes.

3. Minimize the affected area
4. Avoid nested elements with backdrop filters
5. Limit the use of backdrop filters

From [this bug report](https://bugs.chromium.org/p/chromium/issues/detail?id=986206#c59), it still appears to be a Chromium issue and not yet resolved.

**Comment by jasmussen** [comment]
I created a codepen to better explain what I think is going on: https://codepen.io/joen/pen/yLGExva?editors=1100

Here's a GIF:
![color](https://github.com/WordPress/gutenberg/assets/1204802/23c946ae-25c1-4234-874f-f25253605379)

It's showing how when you resize until the blurred area _touches_ the black edge, the black edge starts bleeding into the filter. This is expected. The flickering happens, I believe, when sub-pixel rendering is involved. And we are scaling up the viewport using `transform: scale`, so subpixel handling will be part of it. And so as this happens, as the viewport scales, then every once in a while Chrome's rounding math will switch back and forth between _the black overlapping_ and not.

**Comment by t-hamano** [comment]
Thank you for your detailed explanation, @jasmussen.

In other words, does this mean that this problem may be solved with CSS that has some kind of decimal point value? Sorry if I didn't understand your intentions correctly.

**Comment by jasmussen** [comment]
No, what I mean is that all browsers have some rounding math they do, when calculating a subpixel. Let's say I have a 400px wide div with a blue background. Then I apply a `transform: translateX(0.4px);`. This affects some of the antialiasing of the box. But then if the box is horizontally centered on a 2x resolution screen, that antialiasing will further be nuanced, for every single pixel the window is resized. In some cases it will have to round down, in other cases it will have to round up.

The issue at play is that if this box is set to blur the background, the surroundings sometimes bleed in. This isn't entirely incorrect behavior. It just makes it tricky to cleanly and firmly say that this is a bug in Chrome, which means that it might not ever be fixed on the browser end. I would think that _overflow_ should prevent things from bleeding in, but that's the kind of CSS fixing we might need here.

**Comment by t-hamano** [comment]
Thank you for your detailed explanation.

I've tried several approaches, including using `will-change:transform;` and overflowing the element with blur applied, but I haven't found a solution yet.

By the way, I found a report similar to this issue, so I'll leave it here.

- [Backdrop-filter blur not resizing when changing window size - Stack Overflow](https://stackoverflow.com/questions/75543160/backdrop-filter-blur-not-resizing-when-changing-window-size)
- [Flickering backdrop filter blur on scaling - Stack Overflow](https://stackoverflow.com/questions/73088013/flickering-backdrop-filter-blur-on-scaling)

**Comment by jasmussen** [comment]
Great links, that echoes our conversations. It does suggest it's a chromium bug:

![Screenshot 2023-10-04 at 08 36 39](https://github.com/WordPress/gutenberg/assets/1204802/15a396ef-1772-4aff-a862-30ac3fed445c)

Though i still maintain that it isn't entirely clear that it's a _bug_, because of the nature of how blur should be affected by its surroundings. Here's a simple example in Photoshop of how the blur in the middle of looks like a straight left to right gradient:

![blur 1](https://github.com/WordPress/gutenberg/assets/1204802/5370701e-de5d-4676-8ae6-927f1f1dae03)

This happens because the black square on the left extends all the way from the top to the bottom, therefore the blur in the center can be aware of information above and below.

If the above and below were white, on the other hand, as would the blur be different:

![blur-2](https://github.com/WordPress/gutenberg/assets/1204802/4e164856-da55-423a-8478-c48ad94fb40c)

This is because now there's white information above and below affecting the blur in the middle. This is despite the square in the center not touching the surroundings at all.

That said, if I cropped the canvas to just the square, before the blur, then there would not exist any "outside information" above and below, and we would again get a perfect gradient:

![blur-3](https://github.com/WordPress/gutenberg/assets/1204802/ffc51b34-3002-426d-accf-1259e7b9ad1f)

My main point in illustrating this is, it isn't clear cut whether it's a bug or feature, and for that reason, I don't think Chrome will be making changes anytime soon. Which is mainly an argument that we need an alternative. And just like I suggested, the solution likely means we need _clearance_ around the blurred box, whether through a border or padding, as suggested also here:

![Screenshot 2023-10-04 at 08 36 44](https://github.com/WordPress/gutenberg/assets/1204802/c7f29159-b9ae-4d62-8a11-4c28c17c065b)

I tried to go as low as possible, 0.7, but that may not be enough. I'll try a quick PR.

**Comment by jasmussen** [comment]
Created #55032 which "fixes" this, but the hack is not the greatest. Notably I found more technical debt in that the filter was disabled on the featured image, sort of a throwing of the towel from a previous bout with the same bug. I tried fixing that too, and it got a bit ugly.

The more easy to land version would be an extraction from my PR of just the changes to the generic placeholder component, and then revisit the Featured Image block separately. But have a look, let's see what we can land.

**Comment by jasmussen** [comment]
Just checking in: is this still an issue? My instinct is _yes_, but wondering if anyone has updates.

**Comment by t-hamano** [comment]
In the latest version of Gutenberg, there seems to be no animation when hovering over the canvas. Does anyone know if this is an intentional change or a regression?

Personally, I prefer the latest version without animations.


https://github.com/WordPress/gutenberg/assets/54422211/23db195b-b109-4975-a5b8-e8308ebdf0e0

**Comment by jasmussen** [comment]
@youknowriad do you know? I think you worked on animation recently☝️

**Comment by jasmussen** [comment]
It seems unintentional.

**Comment by youknowriad** [comment]
Yes, I guess one of my refactorings removed the animation, I'm also not sure that the animation adds too much value though.

**Comment by t-hamano** [comment]
How about displaying an outline instead of zooming in on mouse hover? This matches the style when focusing the canvas with the keyboard.

Related PR: https://github.com/WordPress/gutenberg/pull/50296

https://github.com/WordPress/gutenberg/assets/54422211/4eb325c0-bd8f-4628-b643-d45c0efb1248

**Comment by jasmussen** [comment]
> I'm also not sure that the animation adds too much value though.

It feels important to suggest interactivity when clicking on it. As part of the details normalization work, I'd think that's even more important to keep, IMO. But the animation already needed a little work — previously what triggered it was the entire real-estate on the right, as opposed to just _the frame itself_, which meant that even if you had scaled down the preview to a mobile phone size, just hovering any part of the right side of the screen would trigger the behavior.

**Comment by jasmussen** [comment]
I've a tentative flicker fix for the bug that was caused by the resizing in #61201. That of course assumes we'll restore the scaling (which I think we should), and that the way we restore it causes that visual glitch to reapper.

**Comment by t-hamano** [comment]
I checked #61201, but another concern with restoring scaling is that text becomes blurry when hovered. Does this occur in your environment?

The video below was tested with WP6.5 with scaling enabled. You can see more details by viewing the video in full screen.

https://github.com/WordPress/gutenberg/assets/54422211/af413f33-9781-4403-8171-a2161baef9b0

**Comment by jasmussen** [comment]
That does not occur in my env, no. 🤔

**Comment by t-hamano** [comment]
> but another concern with restoring scaling is that text becomes blurry when hovered. 

Is there anyone who can reproduce this problem? You should be able to test it with WordPress6.5.

However, the scaling itself has existed before, so it might be okay to bring it back.

**Comment by annezazu** [comment]
@jordesign can you replicate this still? I wasn't able to replicate in Chrome, 6.5.2, with GB 18.2 installed.

**Comment by jordesign** [comment]
I'm still seeing it in WP 6.5.2 _without_ GB active. But once GB 18.2 is active I confirm I don't see this any more (as the scaling/animation doesn't happen).

**Comment by jasmussen** [comment]
As it turns out, the issue is temporarily fixed due to a regression in trunk. That regression is that the preview pane no longer scales up on mouseover. But this is a scaling that we will likely restore, as it serves a purpose to indicate interactivity. That said, I have a new PR prepped in #61201, that _should_ fix the glitching once the scaling is restored.

**Comment by t-hamano** [comment]
#61647 restored the zoom effect, but for some reason, I can no longer reproduce this problem in my environment (See [this video](https://github.com/WordPress/gutenberg/pull/61647#pullrequestreview-2055056852)). That PR is included in the recently released Gutenberg plugin version 18.4.

Is there anyone who can still reproduce this problem?

**Comment by jasmussen** [comment]
I can also not repro, so tentatively I'm going to close this one. Let's reopen if it it re-emerges.
